packages/graph/hash_graph/lib/graph/src/knowledge/mod.rs

```diff
@@ -4,7 +4,45 @@
 mod entity;
 mod link;
 
+use serde::{Deserialize, Serialize};
+use utoipa::Component;
+
 pub use self::{
     entity::{Entity, EntityId, PersistedEntity, PersistedEntityIdentifier},
     link::Link,
 };
+use crate::{
+    ontology::{
+        PersistedDataType, PersistedEntityType, PersistedLinkType, PersistedPropertyType,
+        QueryDepth,
+    },
+    store::query::Expression,
+};
+
+/// Query to read [`Entities`], which are matching the [`Expression`].
+///
+/// [`Entities`]: Entity
+#[derive(Debug, Deserialize, Component)]
+#[serde(deny_unknown_fields, rename_all = "camelCase")]
+pub struct EntityQuery {
+    #[serde(rename = "query")]
+    pub expression: Expression,
+    #[component(value_type = number)]
+    pub data_type_query_depth: QueryDepth,
+    #[component(value_type = number)]
+    pub property_type_query_depth: QueryDepth,
+    #[component(value_type = number)]
+    pub link_type_query_depth: QueryDepth,
+    #[component(value_type = number)]
+    pub entity_type_query_depth: QueryDepth,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Component)]
+#[serde(rename_all = "camelCase")]
+pub struct EntityRootedSubgraph {
+    pub entity: PersistedEntity,
+    pub referenced_data_types: Vec<PersistedDataType>,
+    pub referenced_property_types: Vec<PersistedPropertyType>,
+    pub referenced_link_types: Vec<PersistedLinkType>,
+    pub referenced_entity_types: Vec<PersistedEntityType>,
```

Comment:
I'm going to put the comment here but I think it applies to this struct, the `EntityQuery` struct, the return values, the querying mechanism, etc.

It's quite weird to me for us to put the type references and depths on the same level as the actual entity thing.

I think we basically want our return value to be of the structure { knowledge: ..., ontology: ...} (to be clear I'm not saying we _should_ use those keys, but I think we want to distinct sections)

Reply:
I think to be consistent, we could change it (in both structs) to have `referenced_entities` (follow-up) on the top level, but the ontology stuff nested in `ontology`. However, I don't know if this a good change, this way we have a flat list of everything. If a user does not require ontology stuff, they could simply pass `0` and only specify the depth for entities:
```
POST http://127.0.0.1:4000/entities/query
Content-Type: application/json

{
  "query": {
    "eq": [
      {
        "path": [
          "version"
        ]
      },
      {
        "literal": "latest"
      }
    ]
  },
  "entityQueryDepth": 1,
  "dataTypeQueryDepth": 0,
  "propertyTypeQueryDepth": 0,
  "linkTypeQueryDepth":0,
  "entityTypeQueryDepth": 0
}
```

In any case, I think the `Query` and the `Subgraph` struct should have the same shape.

Reply:
On GraphQL this would then look like this
```gql
query recursive_entities {
  entity(entityId: "00000000-0000-0000-0000-000000000000") {
    properties
    
    linkedEntities {
    	properties
    }
  }
}
```

Reply:
Just to be pedantic, I don't think we should call it `referenced_entities` as the entities are not referenced. They are linked. (The types _are_ referenced, but they're referenced _by_ the entity type, not by the entity.)